src/container/srv_target.c

```diff
@@ -2393,3 +2394,151 @@ ds_cont_tgt_ec_eph_query_ult(void *data)
+			D_ERROR(DF_UUID" WRITE_DATA capa turned off.\n",
+				DP_UUID(arg->cwd_cont_hdl));
+	} else if (!turn_off && cont_hdl->sch_write_data_off == 1) {
+		capa = cont_hdl->sch_sec_capas;
+		cont_hdl->sch_sec_capas =
+			ds_sec_cont_capa_write_data_enable(capa);
+		cont_hdl->sch_write_data_off = 0;
+		if (dss_get_module_info()->dmi_tgt_id == 0)
+			D_ERROR(DF_UUID" WRITE_DATA capa turned on.\n",
+				DP_UUID(arg->cwd_cont_hdl));
+	}
+
+out:
+	if (cont_hdl != NULL)
+		ds_cont_hdl_put(cont_hdl);
+	return rc;
+}
+
+static int
+cont_rf_check(d_list_t *rlink, void *data)
+{
+	struct cont_rf_check_arg	*arg = data;
+	struct cont_write_data_set_arg	 wd_arg;
+	struct ds_cont_hdl		*cont_hdl;
+	struct ds_cont_child		*cont_child;
+	daos_prop_t			 props;
+	struct daos_prop_entry		 entry;
+	struct daos_co_status		 co_stat;
+	bool				 pool_match;
+	int				 rc = 0;
+
+	cont_hdl = container_of(rlink, struct ds_cont_hdl, sch_entry);
+	cont_child = cont_hdl->sch_cont;
+	if (cont_child == NULL)
+		return 0;
+	pool_match = uuid_compare(cont_child->sc_pool_uuid,
+				  arg->crc_pool->sp_uuid) == 0;
+	if (!pool_match ||
+	    (cont_hdl->sch_flags & DAOS_COO_RF_READONLY) == 0 ||
+	    (!ds_sec_cont_can_write_data(cont_hdl->sch_sec_capas) &&
+	     cont_hdl->sch_write_data_off == 0))
+		return 0;
+
+	props.dpp_nr = 1;
+	props.dpp_entries = &entry;
+	props.dpp_entries[0].dpe_type = DAOS_PROP_CO_STATUS;
+	rc = cont_iv_prop_fetch(arg->crc_pool->sp_iv_ns, cont_child->sc_uuid,
+				&props);
+	if (rc) {
+		D_ERROR("cont_iv_prop_fetch failed, "DF_RC"\n", DP_RC(rc));
+		goto out;
+	}
+	daos_prop_val_2_co_status(entry.dpe_val, &co_stat);
+	rc = ds_pool_rf_verify(arg->crc_pool, co_stat.dcs_pm_ver,
+			       cont_child->sc_props.dcp_redun_fac);
+	if (rc != 0 && rc != -DER_RF)
+		goto out;
+
+	uuid_copy(wd_arg.cwd_cont_hdl, cont_hdl->sch_uuid);
+	wd_arg.cwd_turn_off = (rc == -DER_RF);
+	rc = dss_task_collective(cont_write_data_capa_set, &wd_arg, 0);
+	if (rc)
+		D_ERROR("collective cont_write_data_turn_off failed, "DF_RC"\n",
+			DP_RC(rc));
+
+out:
+	return rc;
+}
+
+static void
+cont_rf_check_ult(void *data)
+{
+	struct cont_rf_check_arg	*arg = data;
+	struct dsm_tls			*tls = dsm_tls_get();
+	int				 rc = 0;
+
+	rc = d_hash_table_traverse(&tls->dt_cont_hdl_hash, cont_rf_check, arg);
+	if (rc)
+		D_ERROR("d_hash_table_traverse, cont_rf_check "DF_RC"\n",
+			DP_RC(rc));
+
+	ABT_eventual_set(arg->crc_eventual, (void *)&rc, sizeof(rc));
+}
+
+/** Check active container handle, it its RF value match with new pool map */
+int
+ds_cont_hdl_rf_check(struct ds_pool *pool)
+{
+	struct cont_rf_check_arg	 check_arg;
+	int				*status;
+	int				 tgt_idx;
+	int				 rc = 0;
+
+	check_arg.crc_pool = pool;
+	rc = ABT_eventual_create(sizeof(*status), &check_arg.crc_eventual);
+	if (rc != ABT_SUCCESS)
+		return dss_abterr2der(rc);
+
+	/* check RF on one random VOS main XS */
+	tgt_idx = rand() % dss_tgt_nr;
+	rc = dss_ult_create(cont_rf_check_ult, &check_arg, DSS_XS_VOS, tgt_idx,
+			    0, NULL);
```

Comment:
if you do collective through update_child_map(), then you do not need create extra ULT here?

btw: why choose an random VOS XS here?  to get container list?    It seems you can do rf_verify in xstream 0, then  set sec_capas on each VOS stream by update_child_map()? or I miss sth?

Reply:
I did not do it in update_child_map() because 1) to keep them relative separate, as one is pool func, another is per container, 2) I want only one XS iterate the active cont hdl, that seems not very convenient to do in update_child_map().
For the extra ULT, looks not a problem as that is only called in pool update?

"why choose an random VOS XS here? to get container list? It seems you can do rf_verify in xstream 0"
only main VOS XS with the active cont hash table created, no? So I picked one random VOS XS to check, and if need set capa will do it by collective.